### PR TITLE
[BYOC][DNNL] Enable layer normalization in DNNL byoc.

### PR DESCRIPTION
**Pull request by billishyahao**
This patch is to enable layer normalization in DNNL BYOC by providing an out-of-box rewrite pattern for combining the operators into a single relay layer normalization operator as well as its implementation in dnnl json codegen.

After applying the rewrite pattern, we will observe the following dnnl function:
```
6202 def @tvmgen_default_dnnl_main_108(%dnnl_108_i0: Tensor[(1, 784, 128), float32], Inline=1, Compiler="dnnl", global_symbol="tvmgen_default_dnnl_main_108", Primitive=1) -> Tensor[(1, 784, 128), float32] {
6203   nn.layer_norm(%dnnl_108_i0, meta[relay.Constant][56] /* ty=Tensor[(128), float32] */, meta[relay.Constant][57] /* ty=Tensor[(128), float32] */) /* ty=Tensor[(1,  784, 128), float32] */
6204 }
```
Once you enable DNNL_VERBOSE flag, more informations are shown in log file as below:

```
onednn_verbose,exec,cpu,layer_normalization,simple_layer_normalization:any,forward_inference,data_f32::blocked:abc:f0 stats_undef::undef::f0 diff_undef::undef::f0,,flags:CH,1x784x128,0.0551758
```

With this patch, I benchmarked the inference performance of a kind of vision-tranformer called PCPVT (https://arxiv.org/abs/2104.13840) on ICX-8352Y. It gains up to 1.18X boost. Here is some boost data:

| 32 cores |Latency |
|--|--|
| baseline byoc | 11.45ms | 
| byoc w/ patch| 9.68ms  |

Thanks for contributing to TVM!   Please refer to guideline https://tvm.apache.org/docs/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

### CODE REVIEW

**Comment by crazydemo** [comment]
Thanks for your contribution for BYOC-DNNL. And my suggestions are listed below:
1. I wonder if we can get better performance via running `layernorm` on `dnnl codegen` than running consecutive ops on `native codegen`. Could you please provide some performance numbers?
2. `Lint` has failed. Please run `task_lint.sh` to check the code style.
3. `UT` is required. You can add your test cases in `tests/python/contrib/test_dnnl.py` to ensure the functionality of the enabled ops.

File: python/tvm/relay/op/contrib/dnnl.py [hunk 1]
```diff
@@ -526,3 +527,52 @@ def visit_call(self, call):
...
+    def __init__(self):
+        super(LayerNormRewrite, self).__init__()
+        self.data = wildcard()
+        self.eps = wildcard()
+        self.gamma = wildcard()
+        self.beta = wildcard()
+        mu = is_op("mean")(self.data)
+        diff = is_op("subtract")(self.data, mu)
+        cdiff = diff | is_op("cast")(diff)
+        p1 = is_op("power")(cdiff, is_constant())
```
Comment (apeskov) [inline python/tvm/relay/op/contrib/dnnl.py:557]: you assume that `is_constant()` is strongly equal 2 but you didn't check that.

File: src/runtime/contrib/dnnl/dnnl_json_runtime.cc [hunk 2]
```diff
@@ -1084,6 +1086,47 @@ class DNNLJSONRuntime : public JSONRuntimeBase {
...
+    auto data_entry = node.GetInputs()[0];
+    auto gamma_entry = node.GetInputs()[1];
+    auto beta_entry = node.GetInputs()[2];
+    
+    dnnl::memory::dims data_shape = nodes_[data_entry.id_].GetOpShape()[data_entry.index_];
+
+    float epsilon = std::stof(node.GetAttr<std::vector<std::string>>("epsilon")[0]);
+
+    // Memory description.
+    dnnl::memory::desc data_md = GenDNNLMemDescByShape(data_shape, dt::f32);
```
Comment (apeskov) [inline src/runtime/contrib/dnnl/dnnl_json_runtime.cc:1101]: You assume that data type is always f32. You should verify that.

File: src/runtime/contrib/dnnl/dnnl_json_runtime.cc [hunk 3]
```diff
@@ -1084,6 +1086,47 @@ class DNNLJSONRuntime : public JSONRuntimeBase {
...
+  void LayerNorm(const size_t& nid) {
+    auto node = nodes_[nid];
+
+    auto data_entry = node.GetInputs()[0];
+    auto gamma_entry = node.GetInputs()[1];
+    auto beta_entry = node.GetInputs()[2];
+    
+    dnnl::memory::dims data_shape = nodes_[data_entry.id_].GetOpShape()[data_entry.index_];
+
+    float epsilon = std::stof(node.GetAttr<std::vector<std::string>>("epsilon")[0]);
```
[the remark this answers is not in the document]
Comment (billishyahao) [reply src/runtime/contrib/dnnl/dnnl_json_runtime.cc]: I add ICHECK to this case. And will update later for supporting all other attributes.

File: src/runtime/contrib/dnnl/dnnl_json_runtime.cc (hunk 2 above)
Comment (billishyahao) [reply src/runtime/contrib/dnnl/dnnl_json_runtime.cc]: The issue is gone after merging with latest main.

**Comment by billishyahao** [comment]
Hi @apeskov , please take a look at the latest version here. Feel free to comment more.

**Comment by billishyahao** [comment]
Hi @masahi Please take a look.

**Comment by billishyahao** [comment]
Hi @comaniac , @trevor-m , @mbaret , Could you take a look at this pr? Thanks!

**Comment by billishyahao** [comment]
> You need to resolve the conflict in `test_dnnl.py`, but since it will be modified in #11513, let's merge #11513 first.

Hi @masahi , Thanks for pointing out this. I have resolved the conflict in test_dnnl.py.

**Pull request merged**